jormungandr/src/network/bootstrap.rs

```diff
@@ -64,44 +61,49 @@ pub async fn bootstrap_from_peer(
             client
                 .pull_blocks_to_tip(&[tip_hash])
                 .map_err(|e| Error::PullRequestFailed { source: e })
-                .and_then(move |stream| bootstrap_from_stream(blockchain, tip, stream, logger))
-        })
-        .and_then(move |tip| {
-            blockchain::process_new_ref(logger2, blockchain2, branch, tip.clone())
-                .map_err(|e| Error::ChainSelectionFailed { source: e })
-                .map(|()| tip)
+                .and_then(move |stream| bootstrap_from_stream(blockchain, branch, tip, stream, logger))
         });
 
     bootstrap.compat().await
 }
 
 fn bootstrap_from_stream<S>(
     blockchain: Blockchain,
+    branch: Tip,
     tip: Arc<Ref>,
     stream: S,
     logger: Logger,
-) -> impl Future<Item = Arc<Ref>, Error = Error>
-where
-    S: Stream<Item = Block, Error = NetworkError>,
-    S::Error: Debug,
+) -> impl Future<Item=Arc<Ref>, Error=Error>
+    where
+        S: Stream<Item=Block, Error=NetworkError>,
+        S::Error: Debug,
 {
-    let fold_logger = logger.clone();
-
     let block0 = blockchain.block0().clone();
+    let logger2 = logger.clone();
+    let blockchain2 = blockchain.clone();
 
     stream
         .map_err(|e| Error::PullStreamFailed { source: e })
         .filter(move |block| block.header.hash() != block0)
-        .fold(tip, move |_, block| {
-            handle_block(blockchain.clone(), block, fold_logger.clone())
+        .and_then(move |block| handle_block(blockchain.clone(), block, logger.clone()))
+        .and_then(move |new_tip| {
+            // A further optimization could be to process bootstrap blocks in batches (e.g. of 100)
+            // and then only call process_new_ref on the last value of each processed batch. However,
+            // at this time, it would seem processing_new_ref for every bootstrap block does not
+            // introduce too much overhead. Thus such an optimization may be premature and/or
+            // unnecessary.
+            blockchain::process_new_ref(logger2.clone(), blockchain2.clone(), branch.clone(), new_tip.clone())
```

Comment:
No need to clone values other than possibly `new_tip`

Reply:
I thought so as well. However, the compiler emphatically disagrees. I think the move semantics are different because this is a stream future? Happy to take advice otherwise.